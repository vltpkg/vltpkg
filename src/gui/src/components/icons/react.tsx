export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
