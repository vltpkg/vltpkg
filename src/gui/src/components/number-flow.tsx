import {
  motion,
  useMotionValue,
  animate,
  AnimatePresence,
} from 'framer-motion'
import { useEffect, useState, forwardRef } from 'react'
import { cn } from '@/lib/utils.ts'
import { tv } from 'tailwind-variants'

import type { Variants, Transition } from 'framer-motion'
import type { VariantProps } from 'tailwind-variants'

const numberFlowVariants = tv({
  slots: {
    wrapper: '',
    p: '',
    span: '',
  },
  variants: {
    variant: {
      default: {
        wrapper: 'inline-flex',
        p: 'overflow-hidden',
        span: 'inline-block text-sm text-current tabular-nums font-mono',
      },
    },
  },
  defaultVariants: {
    variant: 'default',
  },
})

type NumberFlowProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof numberFlowVariants> & {
    start: number
    end: number
    format?: {
      pad?: number
    }
    motionConfig?: {
      duration?: number
      yValue?: number
      blurValue?: number
    }
    classNames?: {
      wrapper?: string
      p?: string
      span?: string
    }
  }

const NumberFlow = forwardRef<HTMLDivElement, NumberFlowProps>(
  (
    {
      className,
      variant,
      start,
      end,
      motionConfig,
      format,
      classNames,
      ...props
    },
    ref,
  ) => {
    const motionValue = useMotionValue<number>(start)
    const [dir, setDir] = useState<'up' | 'down'>('up')
    const [displayNumber, setDisplayNumber] = useState<number>(start)

    const { wrapper, p, span } = numberFlowVariants({ variant })

    const { pad = 2 } = format ?? {}
    const {
      duration = 2,
      yValue = 2,
      blurValue = 1,
    } = motionConfig ?? {}

    const digitVariants: Variants & Transition = {
      initial: (dir: 'up' | 'down') => ({
        y: dir === 'up' ? yValue : -yValue,
        filter: `blur(${blurValue}px)`,
        opacity: 0,
      }),
      animate: {
        y: '0px',
        filter: 'blur(0px)',
        opacity: 1,
      },
      exit: (dir: 'up' | 'down') => ({
        y: dir === 'up' ? -yValue : yValue,
        filter: `blur(${blurValue}px)`,
        opacity: 0,
      }),
      transition: {
        type: 'spring',
        bounce: 0.35,
        duration: 0.3,
      },
    }

    useEffect(() => {
      setDir(end > start ? 'up' : 'down')

      const controls = animate(motionValue, end, {
        duration,
        ease: 'easeInOut',
        onUpdate: latest => {
          setDisplayNumber(Math.round(latest))
        },
      })

      return controls.stop
    }, [start, end, duration, motionValue])

    const paddedDigits = displayNumber
      .toString()
      .padStart(pad, '0')
      .split('')

    return (
      <div
        ref={ref}
        className={cn(wrapper(), classNames?.wrapper, className)}
        data-id="number-flow-wrapper"
        {...props}>
        <p data-id="number-flow-p" className={cn(p(), classNames?.p)}>
          <AnimatePresence mode="popLayout" initial={false}>
            {paddedDigits.map((n, i) => (
              <motion.span
                data-key={`number-flow-span-${n}-${i}`}
                key={`${i}-${n}`}
                custom={dir}
                initial="initial"
                animate="animate"
                exit="exit"
                variants={digitVariants}
                className={cn(span(), classNames?.span)}
                style={{
                  originY: '0px',
                }}>
                {n}
              </motion.span>
            ))}
          </AnimatePresence>
        </p>
      </div>
    )
  },
)

NumberFlow.displayName = 'NumberFlow'

export { NumberFlow }
