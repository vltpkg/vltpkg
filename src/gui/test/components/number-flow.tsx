import React from 'react'
import { test, expect, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { NumberFlow } from '@/components/number-flow.tsx'

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  cleanup()
})

test('number-flow renders default', () => {
  const Container = () => {
    return <NumberFlow start={1} end={1} />
  }

  const { container } = render(<Container />)
  expect(container.innerHTML).toMatchSnapshot()
})

test('number-flow renders with custom classes', () => {
  const Container = () => {
    return (
      <NumberFlow
        start={1}
        end={1}
        classNames={{
          wrapper: 'custom-wrapper',
          p: 'custom-p',
          span: 'custom-span',
        }}
      />
    )
  }

  const { container } = render(<Container />)

  const wrapper = container.querySelector(
    '[data-id="number-flow-wrapper"]',
  )!
  expect(wrapper).toBeTruthy()
  expect(wrapper.className).toContain('custom-wrapper')

  const p = container.querySelector('[data-id="number-flow-p"]')!
  expect(p).toBeTruthy()
  expect(p.className).toContain('custom-p')

  const spans = container.querySelectorAll('span')
  expect(spans.length).toBeGreaterThan(0)
  spans.forEach(span => {
    expect(span.className).toContain('custom-span')
  })

  expect(container.innerHTML).toMatchSnapshot()
})

test('number-flow renders with custom padding', () => {
  const Container = () => {
    return <NumberFlow format={{ pad: 0 }} start={1} end={1} />
  }

  const { container } = render(<Container />)
  expect(container.innerHTML).toMatchSnapshot()
})
