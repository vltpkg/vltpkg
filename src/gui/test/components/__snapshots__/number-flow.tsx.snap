// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`number-flow renders default 1`] = `

<div
  class="inline-flex"
  data-id="number-flow-wrapper"
>
  <p
    data-id="number-flow-p"
    class="overflow-hidden"
  >
    <span
      data-key="number-flow-span-0-0"
      class="inline-block text-sm text-current tabular-nums font-mono"
      style="filter: blur(0px); opacity: 1; transform: none; transform-origin: 50% 0px 0;"
    >
      0
    </span>
    <span
      data-key="number-flow-span-1-1"
      class="inline-block text-sm text-current tabular-nums font-mono"
      style="filter: blur(0px); opacity: 1; transform: none; transform-origin: 50% 0px 0;"
    >
      1
    </span>
  </p>
</div>

`;

exports[`number-flow renders with custom classes 1`] = `

<div
  class="inline-flex custom-wrapper"
  data-id="number-flow-wrapper"
>
  <p
    data-id="number-flow-p"
    class="overflow-hidden custom-p"
  >
    <span
      data-key="number-flow-span-0-0"
      class="inline-block text-sm text-current tabular-nums font-mono custom-span"
      style="filter: blur(0px); opacity: 1; transform: none; transform-origin: 50% 0px 0;"
    >
      0
    </span>
    <span
      data-key="number-flow-span-1-1"
      class="inline-block text-sm text-current tabular-nums font-mono custom-span"
      style="filter: blur(0px); opacity: 1; transform: none; transform-origin: 50% 0px 0;"
    >
      1
    </span>
  </p>
</div>

`;

exports[`number-flow renders with custom padding 1`] = `

<div
  class="inline-flex"
  data-id="number-flow-wrapper"
>
  <p
    data-id="number-flow-p"
    class="overflow-hidden"
  >
    <span
      data-key="number-flow-span-1-0"
      class="inline-block text-sm text-current tabular-nums font-mono"
      style="filter: blur(0px); opacity: 1; transform: none; transform-origin: 50% 0px 0;"
    >
      1
    </span>
  </p>
</div>

`;
